(a) Sample 10 documents from TinyStories and OpenWebText. Using your previously-trained TinySories and OpenWebText tokenizers (10K and 32K vocabulary size, respectively), encode these sampled documents into integer IDs. What is each tokenizer’s compression ratio (bytes/token)?

Deliverable: A one-to-two sentence response.


In [2]:
from pathlib import Path
from cs336_basics.Tokenizer import Tokenizer

tokenizer = Tokenizer.from_files(
    "vocab.json",
    "merges.json",
    special_tokens=["<|endoftext|>"]
)

path = Path("../data/TinyStoriesV2-GPT4-valid.txt")
raw_bytes = path.stat().st_size

token_count = 0
with path.open("r", encoding="utf-8") as f:
    for chunk in f:
        token_count += len(tokenizer.encode(chunk))

print(f"raw_bytes={raw_bytes}")
print(f"token_count={token_count}")
print(f"bytes_per_token={raw_bytes / token_count:.6f}")

raw_bytes=22502601
token_count=5461210
bytes_per_token=4.120442


(b) What happens if you tokenize your OpenWebText sample with the TinyStories tokenizer? Compare the compression ratio and/or qualitatively describe what happens.

Deliverable: A one-to-two sentence response.


In [3]:
from pathlib import Path
from cs336_basics.Tokenizer import Tokenizer

tokenizer = Tokenizer.from_files(
    "vocab.json",
    "merges.json",
    special_tokens=["<|endoftext|>"]
)

path = Path("../data/owt_valid.txt")
raw_bytes = path.stat().st_size

token_count = 0
with path.open("r", encoding="utf-8") as f:
    for chunk in f:
        token_count += len(tokenizer.encode(chunk))

print(f"raw_bytes={raw_bytes}")
print(f"token_count={token_count}")
print(f"bytes_per_token={raw_bytes / token_count:.6f}")

raw_bytes=289998753
token_count=91369966
bytes_per_token=3.173896


(c) Estimate the throughput of your tokenizer (e.g., in bytes/second). How long would it take to tokenize the Pile dataset (825GB of text)?

Deliverable: A one-to-two sentence response.

**Answer**:

- From the earlier owt_valid run I measured 289,998,753 bytes processed in about 137 s, so throughput ≈2.1×10⁶ bytes/s (≈2.1 MB/s).
- The Pile is roughly 825 GB ≈8.9×10¹¹ bytes. At 2.1 MB/s, total time is 8.9×10¹¹ ÷ 2.1×10⁶ ≈4.0×10⁵ s, or about 112 hours—just under 4¾ days of continuous tokenization.
- Even using the slightly faster TinyStories timing (~2.5 MB/s) the job still lands around 4 days, so planning for roughly 4–5 days of wall-clock time is reasonable.


(d) Using your TinyStories and OpenWebText tokenizers, encode the respective training and development datasets into a sequence of integer token IDs. We’ll use this later to train our language model. We recommend serializing the token IDs as a NumPy array of datatype uint16. Why is uint16 an appropriate choice?
